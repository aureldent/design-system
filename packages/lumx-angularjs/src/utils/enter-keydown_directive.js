import { ENTER_KEY_CODE } from '@lumx/core/js/constants';

/////////////////////////////

function EnterKeydownDirective() {
    'ngInject';

    function link(scope, el, attrs) {
        el.on('keydown', (evt) => {
            if (evt.which === ENTER_KEY_CODE) {
                scope.$apply(() => {
                    scope.$eval(attrs.lxEnterKeydown, { $event: evt });
                });

                evt.preventDefault();
            }
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module('lumx.utils.enter-keydown').directive('lxEnterKeydown', EnterKeydownDirective);

/////////////////////////////

export { EnterKeydownDirective };
