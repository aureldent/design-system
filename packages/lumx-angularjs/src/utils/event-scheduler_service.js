function EventSchedulerService($document, LxUtilsService) {
    'ngInject';

    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The handlers.
     *
     * @type {Object}
     */
    const _handlers = {};

    /**
     * The schedule.
     *
     * @type {Object}
     */
    const _schedule = {};

    /////////////////////////////
    //                         //
    //    Private functions    //
    //                         //
    /////////////////////////////

    /**
     * Handle en event.
     *
     * @param {Event} evt The event.
     */
    function _handle(evt) {
        let scheduler;

        for (const [key] of Object.entries(_schedule)) {
            if (key.includes(evt.type)) {
                scheduler = _schedule[key];
            }
        }

        if (angular.isDefined(scheduler)) {
            for (let i = 0, len = scheduler.length; i < len; i++) {
                const handler = scheduler[i];

                if (angular.isDefined(handler) && angular.isDefined(handler.cb) && angular.isFunction(handler.cb)) {
                    handler.cb(evt);

                    if (evt.isPropagationStopped()) {
                        break;
                    }
                }
            }
        }
    }

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Register en event.
     *
     * @param  {string}   eventName The event name.
     * @param  {Function} cb        The event callback.
     * @return {string}   The event id.
     */
    function register(eventName, cb) {
        const handler = {
            cb,
            eventName,
        };

        const id = LxUtilsService.generateUUID();
        _handlers[id] = handler;

        if (angular.isUndefined(_schedule[eventName])) {
            _schedule[eventName] = [];

            $document.on(eventName, _handle);
        }

        _schedule[eventName].unshift(_handlers[id]);

        return id;
    }

    /**
     * Unregister en event.
     *
     * @param  {string} id The event id.
     * @return {found}  Whether the event has been found or not.
     */
    function unregister(id) {
        let found = false;
        const handler = _handlers[id];

        if (angular.isDefined(handler) && angular.isDefined(_schedule[handler.eventName])) {
            const index = _schedule[handler.eventName].indexOf(handler);

            if (angular.isDefined(index) && index > -1) {
                _schedule[handler.eventName].splice(index, 1);

                delete _handlers[id];
                found = true;
            }

            if (_schedule[handler.eventName].length === 0) {
                delete _schedule[handler.eventName];

                $document.off(handler.eventName, _handle);
            }
        }

        return found;
    }

    /////////////////////////////

    service.register = register;
    service.unregister = unregister;
}

/////////////////////////////

angular.module('lumx.utils.event-scheduler').service('LxEventSchedulerService', EventSchedulerService);

/////////////////////////////

export { EventSchedulerService };
