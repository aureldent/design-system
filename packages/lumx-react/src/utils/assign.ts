export const assign = (target: any, ...sources: any[]) => {
    sources.forEach((source) => Object.keys(source).forEach((key) => (target[key] = source[key])));
    return target;
};
