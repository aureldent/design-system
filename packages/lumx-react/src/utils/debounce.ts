export function debounce(func: (...args: any) => any, debounceTime: number) {
    let timeout: any;
    return function debouncedFunction(...args: any[]) {
        clearTimeout(timeout);
        timeout = setTimeout(() => func(...args), debounceTime);
    };
}
