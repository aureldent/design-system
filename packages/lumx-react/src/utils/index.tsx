export { getBasicClass, handleBasicClasses, detectSwipe, onEnterPressed, onEscapePressed } from '@lumx/core/utils';

export * from './calculatePopoverPlacement';

export * from './isInViewPort';

export * from './validation';

export * from './partitionMulti';

export { getRootClassName } from './getRootClassName';

export * from './type';
