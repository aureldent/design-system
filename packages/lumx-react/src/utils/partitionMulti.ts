import { Predicate } from './type';
import reduce from 'lodash/reduce';
import concat from 'lodash/concat';
import dropRight from 'lodash/dropRight';
import partition from 'lodash/partition';
import last from 'lodash/last';

/**
 * Similar to lodash `partition` function but working with multiple predicates.
 *
 * @example
 * const isString = (s) => typeof s === 'string'
 * const isNumber = (s) => typeof s === 'number'
 * const [strings, numbers, others] = partitionMulti(['a', 1, 'b', false], [isString, isNumber])
 * //=> [['a', 'b'], [1], [false]]
 *
 * @param  elements array of elements
 * @param  predicates array of predicates to apply on elements
 * @return partitioned elements by the given predicates
 */
function partitionMulti<T>(elements: T[], predicates: Array<Predicate<T>>): T[][] {
    return reduce(
        predicates,
        (partitioned: T[][], predicate: Predicate<T>) =>
            concat(dropRight(partitioned), partition(last(partitioned), predicate)),
        [elements],
    );
}

export { partitionMulti };
