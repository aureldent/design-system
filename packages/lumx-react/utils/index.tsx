export * from './calculatePopoverPlacement';

export * from './isInViewPort';
