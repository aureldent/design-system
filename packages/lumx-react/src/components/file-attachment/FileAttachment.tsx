import { mdiFile } from '@lumx/icons';
import { COMPONENT_PREFIX } from '@lumx/react/constants';
import { IGenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';
import classNames from 'classnames';
import React, { useCallback, useMemo } from 'react';
import { AspectRatio, ColorPalette, ColorVariant, Size, Theme } from '..';
import { Icon } from '../icon/Icon';
import { Link } from '../link/Link';
import { Thumbnail, ThumbnailVariant } from '../thumbnail/Thumbnail';
import { IFile } from './types';

/////////////////////////////

/**
 * Defines the props of the component.
 */
interface IFileAttachmentProps extends IGenericProps, IFile {
    /** The theme. */
    theme?: Theme;
}
type FileAttachmentProps = IFileAttachmentProps;

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * The display name of the component.
 */
const COMPONENT_NAME = `${COMPONENT_PREFIX}FileAttachment`;

/**
 * The default class name and classes prefix for this component.
 */
const CLASSNAME: string = getRootClassName(COMPONENT_NAME);

/**
 * The default value of props.
 */
const DEFAULT_PROPS: Partial<FileAttachmentProps> = {
    theme: Theme.light,
};

const FileAttachment: React.FC<FileAttachmentProps> = ({
    className,
    icon = DEFAULT_PROPS.icon,
    thumbnail,
    title,
    url = '',
    description,
    theme = DEFAULT_PROPS.theme,
    ...props
}) => {
    const fileName = useMemo(() => url.substr(url.lastIndexOf('/') + 1) || '', [url]);
    const onFileClick = useCallback(() => window.open(url, '_blank'), [url]);
    return (
        <div className={classNames(className, handleBasicClasses({ prefix: CLASSNAME, theme }))} {...props}>
            {thumbnail ? (
                <Thumbnail
                    role="link"
                    variant={ThumbnailVariant.rounded}
                    onClick={onFileClick}
                    tabIndex={0}
                    {...thumbnail}
                    size={Size.l}
                    aspectRatio={AspectRatio.square}
                />
            ) : (
                <Icon
                    role="link"
                    tabIndex={0}
                    onClick={onFileClick}
                    icon={mdiFile}
                    theme={theme}
                    color={theme === Theme.light ? ColorPalette.dark : ColorPalette.light}
                    {...icon}
                    className={classNames(`${CLASSNAME}__icon`, icon?.className)}
                    size={Size.l}
                />
            )}
            <div className={`${CLASSNAME}__infos`}>
                <Link
                    className={`${CLASSNAME}__title`}
                    color={theme === Theme.light ? ColorPalette.dark : ColorPalette.light}
                    colorVariant={ColorVariant.N}
                    href={url}
                >
                    {title}
                </Link>
                <p className={`${CLASSNAME}__description`}>{description}</p>
                <Link
                    className={`${CLASSNAME}__link`}
                    color={theme === Theme.light ? ColorPalette.blue : ColorPalette.light}
                    colorVariant={ColorVariant.N}
                    href={url}
                >
                    {fileName}
                </Link>
            </div>
        </div>
    );
};

FileAttachment.displayName = COMPONENT_NAME;

export { CLASSNAME, DEFAULT_PROPS, FileAttachment, FileAttachmentProps };
