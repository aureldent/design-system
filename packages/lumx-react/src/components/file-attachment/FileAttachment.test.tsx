import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { ICommonSetup, Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { mdiFile, mdiFilePdfBox } from '@lumx/icons';
import { Theme } from '..';
import { CLASSNAME, DEFAULT_PROPS, FileAttachment, FileAttachmentProps } from './FileAttachment';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<FileAttachmentProps>;

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;

    /**
     * The wrapper of the file attachment.
     */
    wrapper: Wrapper;
    /**
     * The thumbnail of the file attachment.
     */
    thumbnail: Wrapper;
    /**
     * The icon of the file attachment.
     */
    icon: Wrapper;
    /**
     * The Link containing the file name.
     */
    fileNameLink: Wrapper;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  props  The props to use to override the default props of the component.
 * @param  [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return An object with the props, the component wrapper and some shortcut to some element inside of the component.
 */
const setup = (props: ISetupProps = {}, shallowRendering: boolean = true): ISetup => {
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    // @ts-ignore
    const wrapper: Wrapper = renderer(<FileAttachment {...props} />);

    return {
        fileNameLink: wrapper.find(`.${CLASSNAME}__link`),
        icon: wrapper.find('Icon'),
        props,
        thumbnail: wrapper.find('Thumbnail'),
        wrapper,
    };
};

describe(`<${FileAttachment.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        // Here are some examples of basic props check.

        it('should use default props', () => {
            const { wrapper, icon } = setup();
            Object.keys(DEFAULT_PROPS).forEach((prop: string) => {
                expect(wrapper).toHaveClassName(
                    getBasicClass({ prefix: CLASSNAME, type: prop, value: DEFAULT_PROPS[prop] }),
                );
            });
            expect(icon).toHaveProp('icon', mdiFile);
        });

        it('should pass the theme to the Icon', () => {
            const expectedTheme = Theme.dark;
            const { icon } = setup({ theme: expectedTheme });
            expect(icon).toHaveProp('theme', expectedTheme);
        });

        it('should pass the icon to the Icon component', () => {
            const expectedIcon = mdiFilePdfBox;
            const { icon } = setup({
                icon: {
                    icon: expectedIcon,
                },
            });
            expect(icon).toHaveProp('icon', expectedIcon);
        });

        it('should display the file name correctly', () => {
            const expectedFileName = 'test.pdf';
            const { fileNameLink } = setup({
                url: 'https://sometest.url/path/to/file/test.pdf',
            });
            expect(fileNameLink).toHaveProp('children', expectedFileName);
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', () => {
        const expectedUrl = 'https://expected.url';
        const { thumbnail } = setup({
            thumbnail: {
                image: 'path/to/test.png',
            },
            url: expectedUrl,
        });
        window.open = jest.fn();
        thumbnail.simulate('click');
        expect(window.open).toHaveBeenCalledWith(expectedUrl, '_blank');
    });
    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        it('should display Thumbnail and no Icon if thumbnail is defined', () => {
            const { icon, thumbnail } = setup({ thumbnail: { image: 'Pouet' } });
            expect(thumbnail).toExist();
            expect(icon).not.toExist();
        });

        it('should display Icon and no Thumbnail if thumbnail is not defined', () => {
            const { icon, thumbnail } = setup({ icon: { icon: mdiFilePdfBox } });
            expect(icon).toExist();
            expect(thumbnail).not.toExist();
        });
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, {}, { className: CLASSNAME });
});
