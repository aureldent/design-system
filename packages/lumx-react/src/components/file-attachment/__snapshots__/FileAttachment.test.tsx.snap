// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileAttachment> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-file-attachment lumx-file-attachment--theme-light"
>
  <Icon
    className="lumx-file-attachment__icon"
    color="dark"
    icon="M13,9V3.5L18.5,9M6,2C4.89,2 4,2.89 4,4V20A2,2 0 0,0 6,22H18A2,2 0 0,0 20,20V8L14,2H6Z"
    onClick={[Function]}
    role="link"
    size="l"
    tabIndex={0}
    theme="light"
  />
  <div
    className="lumx-file-attachment__infos"
  >
    <Link
      className="lumx-file-attachment__title"
      color="dark"
      colorVariant="N"
      href=""
    />
    <p
      className="lumx-file-attachment__description"
    />
    <Link
      className="lumx-file-attachment__link"
      color="blue"
      colorVariant="N"
      href=""
    />
  </div>
</div>
`;
